Improvement of logging for masakari-controller

rkmrHonjo opened the pull request:
This pull-request contains following changes.
- Use oslo.log for logging.
- Add configuration values for logging.
- Improve log messages.

And, remove unnecessary line from masakari-controller/db/api.py

Commit:
Improvement of logging for masakari-controller

* Use oslo.log for logging.
* Add configuration values for logging.
* Improve log messages.

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -697,104 +838,55 @@ def do_instance_evacuate(self, uuid, targethost):
         """
+        logmsg = ("BEGIN msg_with_thread_id: " \
+                  "parameters (rawmsg=%s)" % (rawmsg))
+        LOG.debug(str(threading.current_thread()) + " " + logmsg)
```
IMO, this method shouldn't have BEGIN foo/ END foo message. When you set log level to debug, following messages will be output. I think it's really annoying.

BEGIN msg_with_thread_id: parameters( rawmsg= message_passed_from_caller )
thread_id message_passed_from_caller
END msg_with_thread_id: return message_passed_from_caller

muroi commented:
Thanks the patch!
I'm really surprised at all function has following lines and the method name is hard coded.  If you'd like to output log in begging and ending of methods, we should implement decorator for outputting log and use the decorator. 

```
msg = BEGIN function-name: with arguments
LOG.debug(msg)

msg = END function-name
LOG.debug(msg)
```

muroi commented on masakari-controller/controller/masakari_worker.py:
```diff
@@ -58,9 +64,16 @@ def __init__(self, config_object):
+        msg = "END __init__"
+        LOG.debug(self.rc_util.msg_with_thread_id(msg))
+
     def _get_vm_param(self, uuid):
```
Should this kind of tiny method looks like utility methods have BEGIN/END debug log? It's just call rc_util_api.do_instance_show() and it also output the very similar log. I think it reduces log readability and traceability.

rkmrHonjo commented:
Hi Muroi,

Thank you for your comment.
I fix my code.

Commit:
Use decorator to output "BEGIN"/"END" logs

rkmrHonjo commented:
Hi  @muroi ,

I fixed my code in accordance with your comments.
(If I should squash two commits, please point out.)

P.S.
I didn't remove logs from _get_vm_param().
Your comment is basically right.
But _get_vm_param() will retry do_instance_show() if it is failed.
It doesn't just call rc_util_api.do_instance_show().

muroi commented on masakari-controller/controller/masakari_controller.py:
```diff
@@ -47,10 +46,14 @@
 
+log_process_begin_and_end = LogProcessBeginAndEnd()
+LOG = oslo_logging.getLogger('controller.masakari_controller')
 
```
it's __name__, isn't it?

rkmrHonjo replied:
__name__ is "__main__" in masakari_controller.py .
I want to get logger by "controller.foobar" for applying default log level.
My another idea is to set "controller.masakari_controller" to __name__.
If you have other cool ideas, please talk to me.
(__file__ does not contain "controller".)

muroi replied:
I understand.  In that case, writing the module path directly make sense. I have another idea to enable the file to use __name__ but it will change the init script. I think the change should be in different patch.

I don't think setting 'controller.masakari_controller' to __name__ is good since when we want to use __name__ in other place __name__ doesn't work well.

Another idea is creating a new file that only calls masakari() in the file and init script calls the file. The reason __name__ is set to "__main__" here is the init script for service command calls this file as a command argument directly.

rkmrHonjo replied:
I appreciate your idea!
I will push a patch for resolving it after other problems are fixed.

muroi commented on masakari-controller/controller/masakari_controller.py:
```diff
@@ -47,10 +46,14 @@
 import controller.masakari_worker as worker
 import db.api as dbapi
 
+log_process_begin_and_end = LogProcessBeginAndEnd()
```
If you use instance method as a decorator we should pass logger defined following line (or __name__)  to the instance and use the logger to output messages.  It's hard to read the message since the debug message could be following.

```
controller.masakari_util BEGIN method-name: parameter is () {}
controller.masakari_controller log message1 from method-name
controller.masakari_controller log message2 from method-name
controller.masakari_util END method-name: retturn
```

rkmrHonjo replied:
Your opinion is right.
I try to improve my code.

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -53,6 +51,26 @@
+class LogProcessBeginAndEnd(object):
+    def output_log(self, func):
+        def _output_log(*args, **kwargs):
+            start_msg = ("BEGIN %s: parameters is %s, %s") % (func, args, kwargs)
```
Use func.**name** instead of func.  Specifying 'func' shows method name and memory address. it's no information.

```
# current output:  the address name is not useful
BEGIN <function method-name at 0x7fbb11b4be60>: parameters is (foo) {"key": "value"}
```

s/parameters is/ parameters are/

rkmrHonjo replied:
Thanks, I fix my code.

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -53,6 +51,26 @@
+        def _output_log(*args, **kwargs):
+            start_msg = ("BEGIN %s: parameters is %s, %s") % (func, args, kwargs)
+            LOG.debug(str(threading.current_thread()) + " " + start_msg)
+
```
Use threading.current_thread().getName() instead of current_thread(). It shows thread name and linux time when the thread is generated.

Additionally, if possible, when creating a new Thread(), define thread name with notification id or any id/name related to db row. I think it make us easy to debug.

rkmrHonjo replied:
Thanks, I fix my code and challenge your additional advice.

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -663,52 +702,54 @@ def fetch_servers_on_hypervisor(self, hypervisor):
             raise
 
-
 class RecoveryControllerUtil(object):
```
Just a question. Is there any reason the class is not a sub class of Logger of oslo.logging?  I think we don't need to write long code, not LOG.error(self.rc_util.msg_with_thread_id()) but just LOG.error() , if this class is Logger class and is instantiated as LOG.

rkmrHonjo replied:
There is no reason.
Your idea is better than my code.
I fix my code.

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -718,83 +759,11 @@ class RecoveryControllerUtil(object):
+        :rawmsg : message.
+        :return : message with thread id.
         """
         msg = str(threading.current_thread()) + " " + str(rawmsg)
```
same as above

rkmrHonjo replied:
ditto

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -535,8 +569,10 @@ def __init__(self, config_object):
+                                              connect_retries=api_retries,
+                                              logger=LOG.logger)
 
+    @log_process_begin_and_end.output_log
```
It shouldn't output the log since the password appears in message. or if you really want to output the message here the decorator needs additional codes that can mask the password.

rkmrHonjo replied:
I don't really want to output the message here.
I remove decorator from this method.

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -545,6 +581,7 @@ def _get_session(self, auth_args):
 
         return sess
 
+    @log_process_begin_and_end.output_log
```
same as above

rkmrHonjo replied:
ditto

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -569,89 +606,91 @@ def _fetch_project_id(self):
 
         return projects[0].id
 
+    @log_process_begin_and_end.output_log
```
Why do these methods have the decorator?  When the log level is debug, the decorator shows method name and its arguments and then the message in try-section shows same message again.

rkmrHonjo replied:
On second thought, I think that this decorator is unnecessary.
You are right.
I remove this decorator.
And, I check as to whether there is other unnecessary decorators.

muroi commented on masakari-controller/controller/masakari_worker.py:
```diff
@@ -367,34 +345,30 @@ def _skip_process_accident_vm_recovery(self, uuid, vm_state):
 
         return status
 
+    @log_process_begin_and_end.output_log
```
I have same question about the decorator.  masakari() in masakari_controller.py outputs arguments for the method but the method also is decorated. Do we need same information in log file?

rkmrHonjo replied:
Same as above.

muroi commented:
Thanks for updating. Overall the codes looks good. Great job!

I still have some questions. please check the inline comments.

rkmrHonjo commented:
Thank you for your re-comments.
I check those & fix my code.

muroi commented:
Oops, I forgot commenting an important stuff. The decorator method needs wraps[1] decorator.
1. http://docs.python.jp/3.5/library/functools.html#functools.wraps

rkmrHonjo commented:
Oh...Thank you for pointing out.

Commit:
Fix/Improve of logging

 * Improve the code that adds information of thread to logs.
 * Change module name that is contained in BEGIN/END logs
   from "controller.masakari_util" to caller's module name.
 * Fix missing wraps decorator.
 * Remove unnecessary logs.

rkmrHonjo commented:
Hi @muroi ,

Sorry for taking long to fix my code.
Please check new commit.

muroi commented on masakari-controller/controller/masakari_config.py:
```diff
@@ -179,9 +267,15 @@ def _log_setup(self):
+            logging_context_format_string=logging_context_format_string,
             default_log_levels=logging.get_default_log_levels() +
             ['controller=' + level])
+        RequestContext(thread_name=str(
```
Question. Where the context instance is assigned? _request_store variable holds the thread name and output the name in log, doesn't it.

rkmrHonjo replied:
There are two reasons: 
1. I'd like to make RequestContext holding thread_name and record_identifier for logging.
   So I made the class that is inherited oslo_context.
   (RequestContext in oslo_context does not have instance variables of thread_name/record_identifier.)
2. get_current() of "oslo_context/context.py" will be called when the class that is inherited oslo_context outputs logs.
   The return value of get_current() is the "_request_store" object that is defined in "oslo_context/context.py" .
   That is not the "_request_store" of the class that is inherited oslo_context.
   So I cannot outputs thread_name and record_identifier for logs if I'll just make the class that is inherited oslo_context.
   I overwrite _request_store of oslo_context in masakari_config.py for that reason.

muroi replied:
You want to hold thread name and resource id in the class, right? If so, I understand your purpose.

This RequestContext class needs variables related to keystone, defined in L70-83. I think it's always '-'.

muroi commented on masakari-controller/controller/masakari_starter.py:
```diff
@@ -215,7 +210,8 @@ def add_failed_instance(self, notification_id,
 
         try:
-            db_engine = dbapi.get_engine()
+            self.rc_config.set_record_identifier(notification_id)
```
This looks not a thread safe.  The method sets notification_id to the instance referred by rc_config. But all threads share the instance, so  last spin-up thread overwrites _request_store in masakari_config.py if threads >2 works concurrently.

I guess you introduced the identifier for outputting a thread name in log because of my comment[1]. In that case you can specify a thread name[2] where you generate a new thread[3].  threading.current_thread.getName() returns non default thread name, like Thread-1 or Thread-2, only if you specify a thread name.
1. https://github.com/ntt-sic/masakari/pull/43#discussion_r69078837
2. https://docs.python.org/3/library/threading.html#threading.Thread
3. https://github.com/ntt-sic/masakari/blob/master/masakari-controller/controller/masakari_controller.py#L150

rkmrHonjo replied:
Your comment is maybe right.
I'm checking my codes.

muroi commented on masakari-controller/controller/masakari_config.py:
```diff
@@ -27,10 +27,96 @@
+        if not request_id:
+            request_id = generate_request_id()
+        self.request_id = request_id
+        if overwrite or not get_current():
```
where is get_current() defined?

rkmrHonjo replied:
get_current() is defined in oslo_context.context.
get_current() is called by following stack.

masakari_config.RequestContext.to_dict()
-> oslo_log.formatters.ContextFormatter.format()
   -> oslo_log.formatters.ContextFormatter._update_record_with_context()
      -> oslo_context.context.get_current()

muroi replied:
It should be written here as context.get_current(), shouldn't it?  I think the default value of overwrite variable defined at L61 is True, so no codes doesn't call get_current() and you haven't hit an error, method get_current() is not defined.

muroi commented on masakari-controller/controller/masakari_config.py:
```diff
@@ -27,10 +27,96 @@
+                          copy of the index is not overwritten.
+        """
+        self.thread_name = thread_name
+        self.record_identifier = record_identifier
```
Question: what is the difference between record_identifier and resource_uuid.

rkmrHonjo replied:
record_identifier is the value that is able to specify masakari db record uniquely.
record_identifier value is not always UUID format.
So I didn't set record_identifier values to resource_uuid.

But, I decided to fix my code to include record_identifier values in thread name.
So I remove record_identifier from instance variables.

muroi commented on masakari-controller/controller/masakari_config.py:
```diff
@@ -27,10 +27,96 @@
+    user_idt_format = '{user} {tenant} {domain} {user_domain} {p_domain} ' \
+        + '{thread_name} {record_identifier}'
+
+    def __init__(self, auth_token=None, user=None, tenant=None, domain=None,
```
Doesn't this class need to call super() in **init**()?

rkmrHonjo replied:
Sorry, super() is needed.

muroi commented on masakari-controller/controller/masakari_config.py:
```diff
@@ -27,10 +27,96 @@
+        """
+        self.thread_name = thread_name
+        self.record_identifier = record_identifier
+        self.auth_token = auth_token
```
As I commented, are this kind of variables needed?

rkmrHonjo replied:
This kind of variables become unnecessary by adding super().
I'll remove these variables.

muroi commented on masakari-controller/controller/masakari_config.py:
```diff
@@ -24,8 +24,97 @@
+        self.resource_uuid = resource_uuid
+        self.roles = roles or []
+        if not request_id:
+            request_id = generate_request_id()
```
Completely same method is defined in oslo_context.context. So you don't need to define this method at L42 and you can call the method here as context.generate_request_id().

rkmrHonjo commented:
Hi @muroi ,

I rethought about my code for logging thread name.
Then I understood that my most recent patch is not good.

My fix plan is following:
-  Sorry, I noticed that thread name is already contained in log record object!<br>
  If we specify "%(threadName)s" in logging format string, we'll be able to output thread name to log.
-  So I'll remove the class that is inherited RequextContext.<br>
  It is unnecessary.
- And I'll add "%(threadName)s" to the logging format string configuration.

(Of cource, I haven't forgotten to name thread uniquely.)

Regards.

Commit:
Revoke inheriting RequestContext & Name thread uniquely

 * The class that is inherited RequestContext was revoked.
   It is unnecessary because log record object already has thread name.
 * Thread name became uniquely.
   So the variable that hold record_identifier became unnecessary.

rkmrHonjo commented:
Hi @muroi ,

I pushed a new patch.
Fix points are written in [my previous comment](https://github.com/ntt-sic/masakari/pull/43#issuecomment-235538238).

Regards.

rkmrHonjo commented:
Oops, some useless/bad codes might remain...
I'm terribly sorry, please wait until the end of my re-checking.

Commit:
Make log message format configurable

rkmrHonjo commented:
Dear @muroi ,

Sorry to keep you wainting.
Re-checking by myself was done.
Please review my codes.

muroi commented:
thank updating, @rkmrHonjo . LGTM, waiting others comments.

sampathP commented:
@rkmrHonjo Thanks for fix.
@muroi Thanks for review.
LGTM, good for merge.
